### DISCUSSION TITLE
HAML linter available but not enabled

### DISCUSSION THREAD

**Discussion by kleinjm** [w0rp/ale#4309] [Q&A - Ask for help with problems] [answered]
I'm using https://github.com/sds/haml-lint to lint my haml files however, I can't get the violations to appear via Ale.

ALEInfo
```
 Current Filetype: haml
Available Linters: ['hamllint']
  Enabled Linters: []
  Ignored Linters: []
 Suggested Fixers:
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
 Linter Variables:
 Global Variables:
  ...
```

It appears the linter is available. `:! haml-lint` is able to execute in Vim. It seems that I somehow just need to enable the linter with an ale config? I've tried the following different options.

```
let g:ale_linters = {'haml': 'haml-lint'}
let g:ale_linters = {'haml': ['haml-lint']}
let g:ale_linters = {'haml': 'hamllint'}
let g:ale_linters = {'haml': ['hamllint']}
let g:ale_linters = {'haml': 'all'}
```

**Comment by hsanson** [comment, marked as answer]
Try inspecting the `g:ale_linters` within vim. Maybe something is overriding it or you are using a different configuration file.  I tested on my machine and setting `let g:ale_linters = {'haml': ['hamllint']}` properly shows the linter enabled:

```
 Current Filetype: haml
Available Linters: ['hamllint']
  Enabled Linters: ['hamllint']
  Ignored Linters: []
 Suggested Fixers:
  'remove_trailing_lines' - Remove all blank lines at the end of a file.
  'trim_whitespace' - Remove all trailing whitespace characters at the end of every line.
........ cut ...........
let g:ale_lint_on_enter = 1
let g:ale_lint_on_filetype_changed = 1
let g:ale_lint_on_insert_leave = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'normal'
let g:ale_linter_aliases = {'asciidoctor': 'asciidoc'}
let g:ale_linters = {'haml': ['hamllint']}
let g:ale_linters_explicit = 0
let g:ale_linters_ignore = {}
let g:ale_list_vertical = 0
```

To check the variable inside vim you can use `:echo g:ale_linters`.  You may also want to check `b:ale_linters`, `g:ale_linters_ignore`, `b:ale_linters_ignore`

**Reply by kleinjm** [reply]
Thanks for the quick reply!

Before setting `let g:ale_linters = {'haml': ['hamllint']}`, inspecting `g:ale_linters` shows `{}`. Setting `let g:ale_linters = {'haml': ['hamllint']}` and reloading my vim config shows `{'haml': ['hamllint']}` which seems correct. However, I'm still seeing `Enabled Linters: []` with `ALEInfo`.

The other output you recommended shows the following:
`b:ale_linters` = `['brakeman', 'rails_best_practices', 'rubocop', 'ruby']`
`g:ale_linters_ignore` = `{}`
`b:ale_linters_ignore` = `E121: Undefined variable: b:ale_linters_ignore`

Any other ideas? Thanks again.

**Reply by hsanson** [reply]
All I can say is that your configuration needs some cleaning.  I recommend you create a minimal vimrc first and start adding options to see what may be causing issues.

For starters `g:ale_linters` must be set before ALE starts so it must be set in your vimrc. Second that `b:ale_linters` is broken since it is an array when it should be a dictionary. Since it takes precedence over `g:ale_linters` it probably is causing the issue. I would try to figure out what is setting it and remove it.

**Reply by kleinjm** [reply]
https://github.com/kleinjm/dotfiles/blob/9c4e7302844654690295e253067e99ad89bd776c/shared/vim/.vim/ftplugin/ruby.vim#L2

I had a file type plugin setup for ruby that was misconfigured and overwriting the default ALE config. Forgot I set that up 4 years ago 🙈 . Thanks for the pointers!

P.S. I'm sure that array config isn't the correct setup but it has been working well for `.rb` files for years 🤷‍♂️
